subtests/docker_cli/liverestore/liverestore.py

```diff
@@ -0,0 +1,200 @@
+        """
+        Start a container. We only need it briefly, until we (the test,
+        running in host space) can cd into its root filesystem. Container
+        will spin idly until a semaphore file is removed.
+        """
+        c_name = self.stuff['dc'].get_unique_name()
+
+        fin = DockerImage.full_name_from_defaults(self.config)
+        subargs = ['--detach', '--name=' + c_name, fin,
+                   'bash -c "echo READY;'
+                   ' while :; do date +%s; sleep 1;done"']
+        dkrcmd = AsyncDockerCmd(self, 'run', subargs)
+        dkrcmd.execute()
+        dkrcmd.wait_for_ready(c_name)
+        self.stuff['container_name'] = c_name
+
+    def _verify_that_date_changes(self):
+        """
+        Verify that container process is running, by running 'docker log'
+        on it sequentially and making sure that (1) we get new output,
+        and (2) the new output corresponds to current time.
+        """
+        log_0 = self._container_log()
+
+        # Wait up to 5 seconds for output to change
+        end_at = time.time() + 5
+        while time.time() <= end_at:
+            log_1 = self._container_log()
+            if log_1 != log_0:
+                break
+            time.sleep(1)
+
+        self.failif(len(log_0) == len(log_1),
+                    "Container log line count did not grow")
+
+    def _container_log(self):
+        """
+        Returns container log as a list of strings
+        """
+        c_name = self.stuff['container_name']
+        result = mustpass(DockerCmd(self, "logs", [c_name]).execute())
+        return result.stdout.strip().split("\n")
+
+    def _container_pid(self):
+        dc = self.stuff['dc']
+        inspect = dc.json_by_name(self.stuff['container_name'])
+        return int(inspect[0]['State']['Pid'])
+
+    def _container_is_child_of_docker_daemon(self):
+        """
+        Returns True if our container's PID is a child of the docker daemon.
+        """
+        dockerd_pid = docker_daemon.pid()
+
+        pid = self._container_pid()
+        while pid != 1:
+            pid = self._ppid(pid)
+            if pid == dockerd_pid:
+                return True
+        return False
+
+    @staticmethod
+    def _ppid(pid):
+        """ returns parent PID of given PID """
+        return int(utils.run('ps -o ppid= -p %d' % pid).stdout.strip())
+
+    def postprocess(self):
+        super(liverestore, self).postprocess()
+
+        # Run 'docker ps', confirm that container is listed and is active
+        dc = self.stuff['dc']
+        docker_ps = dc.list_containers_with_name(self.stuff['container_name'])
+        self.failif(not docker_ps,
+                    "Container not in 'docker ps' output after restart")
+        self.failif_not_in("Up ", docker_ps[0].status,
+                           "container status as reported by 'docker ps'")
+
+        # Confirm that container is still running (spitting out dates)
+        self._verify_that_date_changes()
```

Comment:
Hmmm, just occurred to me a simpler and faster state-full liveliness test could be:

#. Record the contents of ``/proc/<pid>/stat``.
#. Stop / start daemon
#. Compare before/after ``stat`` fields 22 and 14:   ``starttime_before == starttime_after && utime_before < utime_after``

The you only need the busy-loop running inside the container, and don't need logging or timeout-loop or date checks at all.  It's basically impossible for the process PID to remain the same, and also have the starttime match and utime advance (CPU ticks).

Up to you, but I think it would be simpler.